flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -3853,6 +3853,7 @@ class ArrayExprLowering {
   // Lower the expression in a scalar context.
   template <typename A>
   ExtValue asScalar(const A &x) {
+    Fortran::lower::StatementContext stmtCtx;
```

Comment:
I'm a little uneasy about creating a new statement context in the middle of an expression. I'd like to understand the implications of that with respect to transitions in rank in the subexpressions in an expression.

For instance, if we have something like
```fortran
subroutine p(a)
  integer :: a(10,10)
  interface
    function f2(n)
      integer n
      integer f2(10,10)
    end
  end interface
  interface
    function f0a(n)
      integer n(10)
      integer f0a
    end
  end interface
  interface
    function f0b(n)
      integer n(10)
      integer f0b
    end
  end interface
  interface
    function f1(n)
      integer n
      integer f1(10)
    end
  end interface
  a = f2(f0a(f1(f0b([1,2,3,4,5,6,7,8,9,0]))))
end
```
This transitions from ranked -> scalar -> ranked -> scalar -> ranked, so we want to be sure the temporaries created for the ranked results are cleaned up and cleaned up after they are dead, etc.

Reply:
Here is a test case where a number of compilers, including f18 with and without this fix, generate the same result, as given.  The high level fir for program p has 33 `fir.freemem` ops.  Without this fix, the 33 `fir.freemen` ops are clustered together in a tight sequence near the end of the function.  With this fix, clusters of 2, 3, 8, and 10 `fir.freemem` ops appear earlier, with the remaining 10 ops  at the end of the function.

Output
```
 2 3 1 5 6 3 3 3 6 6
 4 6 6 3 5 10 12 9 9 12 12 24 30
 6 9 9 9 6 5 10 15 18 18 18 18 27 27 54 81
```
Program
```
function s2s(x) ! scalar -> scalar
  integer x
  integer s2s
  s2s = 2 * x
end

function s2a(x,n) ! scalar -> array
  integer x
  integer n
  integer s2a(n)
  s2a = 3 * x
end

function a2s(x,n) ! array -> scalar
  integer x(n)
  integer n
  integer a2s
  a2s = sum(x)
end

function a2a(x,n) ! array -> array
  integer x(n)
  integer n
  integer a2a(n)
  a2a = 5 * x
end

program p
  interface
    function s2s(x)
      integer x
      integer s2s
    end
  end interface

  interface
    function s2a(x,n)
      integer x
      integer n
      integer s2a(n)
    end
  end interface

  interface
    function a2s(x,n)
      integer x(n)
      integer n
      integer a2s
    end
  end interface

  interface
    function a2a(x,n)
      integer x(n)
      integer n
      integer a2a(n)
    end
  end interface

  do i = 1,3
    print*, [ s2s(i) &
            , s2a(i,i) &
            , a2s([(j,j=1,i)],i) &
            , a2a([(j,j=1,i)],i) &
            , s2a(s2s(i),1) &
            , s2a(a2s([(j,j=1,i)],i),i) &
            , a2s(s2a(i,i),i) &
            , a2s([s2a(i,i)],i) &
            , a2s([[s2a(i,i)],[s2a(i,i)]],2*i) &
            , a2s([[s2a(a2s([(j,j=1,i)],i),i)],[s2a(i,i)]],2*i) &
            ]
  enddo
end
```

Reply:
I share Eric's concern here because:
- It may rightfully be used to deal with character scalar expressions that may be temps (see TRIM results) and should not be cleaned-up before their usage.
- I know that `asScalar` is or has been used to lower some arrays with the scalar lowering framework, so I am not entirely sure it always deals with scalar values, in which case it would be bad if it returned an address for something it cleaned-up. Again, not sure about that point, and we should probably find a better name if this still happens because `asScalar` may be a bit misleading here. I'll try to double check that point.

So I think you might want to add an assert that `asScalar` is never returning an address/box here and run it on a full test suit. I am pretty confident it might fire. Ideally, it would fire in the lit tests already.

Reply:
The lowering of some array constants may still be done by "scalar" expression lowering.